pilot/pkg/model/jwks_resolver.go

```diff
@@ -302,42 +326,44 @@ func (r *jwksResolver) refresh(t time.Time) {
 		jwksURI := key.(string)
 		e := value.(jwtPubKeyEntry)
 
-		// Remove cached item if it hasn't been used for a while.
-		if now.Sub(e.lastUsedTime) >= r.evictionDuration {
+		// Remove cached item for either of the following 2 situations
+		// 1) it hasn't been used for a while
+		// 2) it hasn't been refreshed successfully for a while
```

Comment:
If there is a transport error and we can't fetch a new JWT token, then I think we should continue using the old token until connectivity is restored. 

But there may be a tough tradeoff here:  security vs. availability.    If we continue to use the old token forever, will it eventually expire?  Will we compromise security because we can't revoke it and because the token will take too long to expire???  Or are these checks handled by the peer anyways?

If there aren't security concerns about using an old token, then I think it's an easy decision:  keep using cached tokens if you can't fetch new ones due to transport errors.

@wattli @liminw @rshriram @rlenglet @fpesce would love your advice

Reply:
I think there is security concern if we never refresh the public key which might have been deprecated by the idp provider. Typically public key doesn't change very often though, I think one week sounds a reasonable default.

Can we have a follow-up PR to report metrics about public key refresh failure? that would give customers enough time to fix some network and idp setup issue. Also, is this configurable? if not, can we have another follow-up pR to make this (how long the old public key will be kept) configurable? customers can set their own default based on their need.

Reply:
Yeah, one possible situation is the idp stops their service but the customer doesn't know it.

Will add the metric in a follow up PR.

I believe @pitlv2109 has a PR for making these configurable some time ago, but it's de-prioritized at that time.

Reply:
+1 to add metrics to report key refresh failure, and make the cache duration configurable. 

The keys in the cache will have to expire at some point. Keeping the old public keys in the cache for too long without customer knowing it sounds like a security hole to me. If there is prolonged (over a week, based on the PR) network issue to refresh the public keys, traffic interruption (authentication failure) is expected. We just need to make sure that we recover fast when network issue is fixed.

Reply:
@costinm said Istio is moving away from using and/or introducing more command line flags. I believe we talked about introducing a security CRD where we put this type of configs like cache expiration? @liminw @yangminzhu

Reply:
@pitlv2109 somewhat related doc on this https://docs.google.com/document/d/1M-qqBMNbhbAxl3S_8qQfaeOLAiRqSBpSgfWebFBRuu8/edit?ts=5cd35cdf#heading=h.qex63c29z2to

Reply:
@pitlv2109 The other (faster) approach is to use environment variables, but might be a little hard for use to configure.

Reply:
@duderino @rshriram @rlenglet @fpesce 
I have tested the PR in a cluster with a mock jwksUri server simulating the network error, and my test application is working as expected in the cluster since last Friday. Could you let me know if you have any other comments? Thank you.

Reply:
> But there may be a tough tradeoff here: security vs. availability. If we continue to use the old token forever, will it eventually expire? Will we compromise security because we can't revoke it and because the token will take too long to expire??? Or are these checks handled by the peer anyways?

@duderino 
My 2 cents:
If I understand the change correctly it seems OK to me, depending on the evictionDuration default obviously. A whole week seems a bit loose.
Also, if I understand that code correctly, there is no story for certificate revocation, i.e.: we might want a way to actively evict a pubkey from cache in case of key compromission, that would mean a way to trigger a call to the refresh() function, other than just the periodical refresher / ticker one here.

Reply:
> i.e.: we might want a way to actively evict a pubkey from cache in case of key compromission, that would mean a way to trigger a call to the refresh() function, other than just the periodical refresher / ticker one here.

One way to do this is users to remove the jwksUri with from their authentication policy, this will force pilot to envoy immediately.

Reply:
@yangminzhu So, if I understand correctly, to revoke a pubkey, we need to remove the jwksUri, and then add it again? (because, just removing the jwksUri will defeat the purpose right? As it is the place the pubkey is loaded from, if you remove it, no other pubkey is available)

Reply:
Right, that's the way to force it to sync immediately, the user could also wait for 20 mins to let it to rotate automatically.

Reply:
@duderino @fpesce It seems there is no other concern about the current PR, if so can we merge this PR? Thanks.